CMSgov/easi-app
# Send fallback email to GRT when an intake has an invalid or missing EUA ID

# EASI-1569

## Changes and Description

* When a system intake has an invalid or missing EUA ID, send a fallback email to the governance review team.
** Missing EUA IDs are `NULL` in the `system_intakes.eua_user_id` column. As far as I can tell, these only occur with intakes that were imported from Sharepoint. Attempting to call the CEDAR LDAP API in this case yields a 500 errors. 
** Invalid EUA IDs appear to correspond to CMS employees/contractors who are no longer at CMS. Calling the CEDAR LDAP API with such an ID returns a 200 OK response with blank fields for all user info:
```
{
    "givenName": "",
    "surName": "",
    "commonName": "",
    "telephoneNumber": "",
    "email": "",
    "userName": ""
}
```
** Regardless of the action taken on the intake request, the same template is used whenever an EUA ID is invalid, and a second template is used whenever an EUA ID is missing. These email templates mention the project name, the EUA ID (if invalid), the request ID, and link to the GRT view of the intake request.
** I went through and looked for all calls to email methods that can send mail to an EUA ID other than the current user; the only four I found were 1 in `pkg/services/action.go` (in `NewCreateActionUpdateStatus()`), and 3 in `pkg/services/system_intakes.go` (in `NewProvideGRTFeedback()`, `NewUpdateLifecycleFields()`, and `NewUpdateRejectionFields`).

## How to test this change
* There are unit tests for the code I added to the `services` package. These can be run with the following commands, each run separately:
```
go test "github.com/cmsgov/easi-app/pkg/services" -testify.m TestCreateActionUpdateStatus
go test "github.com/cmsgov/easi-app/pkg/services" -testify.m TestProvideGRTFeedback
go test "github.com/cmsgov/easi-app/pkg/services" -testify.m TestUpdateLifecycleFields
go test "github.com/cmsgov/easi-app/pkg/services" -testify.m TestUpdateRejectionFields
```
These unit tests require the database container to be up and running (due to how the test suite is set up), but these particular tests don't interact with the database.
* Additionally, I added a container running [MailCatcher](https://mailcatcher.me/) to our Docker Compose setup, which provides a basic mail server running on port 1025 and a web UI for viewing all sent mail at `http://localhost:1080/`. Bringing the app up locally with `scripts/dev up` will also run the MailCatcher container and configure the backend to use it as a mail server. ** To test with a `NULL` EUA ID, create an intake as a user, manually set `eua_user_id` in the database to `NULL`, then provide GRT feedback/issue or extend an LCID/issue a rejection/take another action. You should see the notification email for a missing EUA ID appear in MailCatcher, with a valid link to the intake.
** I can't figure out a way to test for an invalid EUA ID in the live app (which is what spurred writing the unit tests).
* I added a utility script, `cmd/test_email_templates/main.go`, which directly calls the new email client methods. With the backend up, run `go run cmd/test_email_templates/main.go`, then emails for an invalid EUA ID and a missing EUA ID should be available in MailCatcher. The links won't work, though, since the IDs are random.

## Areas for Potential Feedback
* The general code structure in `pkg/services/action.go` and `pkg/services/system_intakes.go` feels a bit clunky and convoluted to me; if there's a way to simplify it, I'm definitely open to feedback.

## Miscellaneous Notes
* The check for an empty EUA ID in `pkg/cedar/cedarldap/translated_client.go` should never be triggered now, but if another code path does call `FetchUserInfo()` with an empty EUA ID in the future, we'll get a more specific error report than the 500 error from calling CEDAR.


Notification for an invalid EUA ID:
![euaID notification mail - invalid](https://user-images.githubusercontent.com/92891039/158685571-552551cd-d793-4c7e-8738-04d622302d74.PNG)

Notification for a missing EUA ID:
![euaID notification mail - missing](https://user-images.githubusercontent.com/92891039/158685611-2a4aa46f-d1d2-491f-956a-e8f3f819438e.PNG)
